### ISSUE TITLE
Understanding the source code

### ISSUE THREAD

**Issue by fullofcaffeine** [skial/haxe.io#97]
Hey Skial,

I was curious about how the web app for Haxe.io was implemented and started digging the code. At first I thought uFront was used, but doesn't look like it. 

But what called my attention was the Main.hx file. Why is its body almost all commented out?

**Comment by skial** [comment]
Hi @fullofcaffeine,

Unfortunately your timing to look into how haxe.io works is bad, as last weekend I had an issue which required me to rebuild all the plugins, but I unfortunately failed to keep the dependencies in check. Basically, you are extremely unlikely to build haxe.io with everything in its current state.

If your interested in how it used to work and its current state, more in-depth detail below.

##### How it _used_ to work.

The main program which handles all the file discovery is [Tuli](https://github.com/skial/tuli) which consists of [LibRunner.hx](https://github.com/skial/tuli/blob/6eaf87d73546dd0c28a2db53f0b6ff92402a1ce9/src/LibRunner.hx) and [Tuli.hx](https://github.com/skial/tuli/blob/6eaf87d73546dd0c28a2db53f0b6ff92402a1ce9/src/uhx/sys/Tuli.hx). `LibRunner.hx` is the entry point from the haxelib command `haxelib run tuli`. 
+ `LibRunner.hx` depends on the [cmd](https://github.com/skial/cmd) library which consists of [Lod](https://github.com/skial/cmd/blob/378774cdb43fcb0c2a5cc0016a13c6666496ca7b/src/uhx/sys/Lod.hx), a command line arguments parser, [Liy](https://github.com/skial/cmd/blob/378774cdb43fcb0c2a5cc0016a13c6666496ca7b/src/uhx/sys/Liy.hx), which at runtime using reflection attempts to map the commands to methods and properties and finally [Ede](https://github.com/skial/cmd/blob/378774cdb43fcb0c2a5cc0016a13c6666496ca7b/src/uhx/sys/Ede.hx), which does the same thing as `Liy` but as a build macro.
+ `Tuli.hx` depends on the [Tappi](https://github.com/skial/tappi) library which finds and loads other `neko` files as plugins.

Tuli reads a [config.json](https://github.com/skial/haxe.io/blob/6e8245594ed5a52f0966d059cdd3ad69ccc420a6/config.json) file located in the root directory, which [lists the plugins](https://github.com/skial/haxe.io/blob/6e8245594ed5a52f0966d059cdd3ad69ccc420a6/config.json#L61) needed to build the final static output. Tuli has a [bunch of plugins included](https://github.com/skial/tuli/tree/master/src/uhx/tuli/plugins) which are built as individual `neko` files. This repo also includes some site specific [plugins](https://github.com/skial/haxe.io/blob/6e8245594ed5a52f0966d059cdd3ad69ccc420a6/plugins.hxml).

I won't go into detail on how _each_ plugin works, but the basis is that each plugins entry point, usual `static function main() {..}` returns itself which is then stored by `Tappi`. When `Tuli` gets the map of loaded plugins, it creates an instance of each class and passes to the constructor an instance of itself. In the plugins constructor it can register callbacks with `Tuli`.

`Tuli` allows plugins to register to run on every file found by `Tuli`, to run on a specific file extension, to run on any data loaded and to run when `Tuli` is about to finish.

If any file content was read, it will be cached by `Tuli`, also if any content was generated by a plugin it also is cached by `Tuli`, all of which was then saved to the `output` directory in the `config.json` file.

##### Current refactor

Because of the issues I've had last weekend, I'm rebuilding each plugin as a separate, individual command line tool so no dependency on `Tappi` and no dependency on `Tuli`, the Haxe community really doesn't need another build tool.

So far the `markdown` and `codehighlighter` plugins now run as standalone command line tools. I'm currently working on [fisel](https://github.com/skial/fisel) which will be an enhanced version of the `ImportHTML` plugin but as a command line tool. I plan on integrating this into [Tem](https://github.com/skial/uhu/issues/20) in the distant future :wink:.

If you have any other questions, I'll happy to answer them.

**Comment by fullofcaffeine** [comment]
Interesting, thanks for taking the time to explain. I now see it's a preprocessor/generator.

I will close this one and wait for the updates. Maube this could be extracted into a preprocessor framework for Haxe, like Jekyll/Middleman is for Ruby?

**Comment by skial** [comment]
Yep, thats what [Tuli](https://github.com/skial/tuli) was initial created for but its [similar to metalsmith](http://www.metalsmith.io/#a-little-secret) as it only handles loading plugins and files and then passing files around to those plugins.

My problem came from the `ImportHTML` plugin, which was caused by my [fork](https://github.com/skial/detox/tree/separate_platforms) of [Detox](https://github.com/jasononeil/detox) by @jasononeil, which uses the HTML parser in [Mo](https://github.com/skial/mo) on `sys` platforms. Detox by default uses the Xml parser from the std library which is just too strict.

So Tuli works well enough to load plugins, pass files around and file creation, as long as each plugin follows this [format](https://github.com/skial/tuli/blob/6eaf87d73546dd0c28a2db53f0b6ff92402a1ce9/src/uhx/tuli/plugins/NoFollow.hx) shown between lines 17 and 24 and compiled to a `neko` file.

I'll try and remember to post to this thread when things get more stable/usable :smile:.
